### PR TITLE
133 stream test harness

### PR DESCRIPTION
**Pull request by Schmed**
I'm still getting the following failure, but I believe I get the same failure without modifying master:

CrawlTopologyTest.testAsync:136 » Runtime Job 'flink-crawler' did not terminat...

### CODE REVIEW

File: src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseProcessFunction.java [hunk 1]
```diff
@@ -11,23 +11,25 @@
 public abstract class BaseProcessFunction<IN, OUT> extends ProcessFunction<IN, OUT> {
 
     protected transient int _parallelism;
-    protected transient int _partition;
-
+    protected transient int _operatorIndex;
```
Comment (Schmed) [inline src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseProcessFunction.java]: I could change this back to _partition if you like, but I was just trying to play along with the names used by the streaming test harness.

**Comment by kkrugler** [comment]
Hi @Schmed - if you re-merge master, I'll do a final review and then approve/merge it.

File: pom.xml [hunk 2]
```diff
@@ -86,7 +94,21 @@
 
             <dependency>
                 <groupId>org.apache.flink</groupId>
-                <artifactId>flink-hadoop-compatibility_2.11</artifactId>
+                <artifactId>flink-test-utils-junit</artifactId>
+                <version>${flink.version}</version>
```
Comment (kkrugler) [inline pom.xml]: Should this be scope test?

File: src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseProcessFunction.java (hunk 1 above)
Comment (kkrugler) [reply src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseProcessFunction.java]: I think `_operatorIndex` makes more sense, but then we should use that everywhere else.

File: src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseCoFlatMapFunction.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.functions.co.RichCoFlatMapFunction;
+
+import com.scaleunlimited.flinkcrawler.pojos.BaseUrl;
+import com.scaleunlimited.flinkcrawler.utils.UrlLogger;
+
+@SuppressWarnings("serial")
+public abstract class BaseCoFlatMapFunction<IN1, IN2, OUT> 
+	extends RichCoFlatMapFunction<IN1, IN2, OUT> {
```
Comment (kkrugler) [inline src/main/java/com/scaleunlimited/flinkcrawler/functions/BaseCoFlatMapFunction.java]: Why is this base class needed?

File: src/main/java/com/scaleunlimited/flinkcrawler/functions/TimerFunction.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
+import org.apache.flink.util.Collector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("serial")
+public class TimerFunction
+    extends KeyedProcessFunction<Integer, Integer, Integer> {
+    static final Logger LOGGER = LoggerFactory.getLogger(TimerFunction.class);
```
Comment (kkrugler) [inline src/main/java/com/scaleunlimited/flinkcrawler/functions/TimerFunction.java]: Javadoc on why this class is needed?

Comment (kkrugler) [reply src/main/java/com/scaleunlimited/flinkcrawler/functions/TimerFunction.java]: also, if this is only used for tests, should it be in `src/test/java`?

File: src/main/java/com/scaleunlimited/flinkcrawler/tools/TimerTool.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.streaming.api.functions.source.SourceFunction;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.scaleunlimited.flinkcrawler.functions.TimerFunction;
+
+public class TimerTool {
```
Comment (kkrugler) [inline src/main/java/com/scaleunlimited/flinkcrawler/tools/TimerTool.java]: Javadoc on why this is needed. Also, if only for testing then maybe in `src/test/java`?

File: src/test/java/com/scaleunlimited/flinkcrawler/functions/TimerTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.flink.streaming.util.KeyedOneInputStreamOperatorTestHarness;
+import org.apache.flink.streaming.util.OneInputStreamOperatorTestHarness;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.scaleunlimited.flinkcrawler.tools.TimerTool;
+
+public class TimerTest {
```
Comment (kkrugler) [inline src/test/java/com/scaleunlimited/flinkcrawler/functions/TimerTest.java]: This is just to validate that timers get saved with state, yes? Do we still want/need it?

**Pull request merged**